File: src/main/java/net/x3djsonld/data/ScubaTank.java
package net.x3djsonld.data;

import org.web3d.x3d.jsail.Core.*;
import org.web3d.x3d.jsail.EnvironmentalEffects.*;
import org.web3d.x3d.jsail.fields.*;
import org.web3d.x3d.jsail.Geometry3D.*;
import org.web3d.x3d.jsail.Grouping.*;
import org.web3d.x3d.jsail.Shape.*;

// Javadoc metadata annotations follow, see below for X3DJSAIL Java source code.
/**
 * <p> Scuba gear used by Nancy Diving example. </p>
 <p> Related links: ScubaTank.java source, <a href="https://www.web3d.org/x3d/content/examples/X3dResources.html" target="_blank">X3D Resources</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dSceneAuthoringHints.html" target="_blank">X3D Scene Authoring Hints</a>, and <a href="https://www.web3d.org/x3d/content/X3dTooltips.html" target="_blank">X3D Tooltips</a>. </p>
	<table style="color:black; border:0px solid; border-spacing:10px 0px;">
        <caption>Scene Meta Information</caption>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center; padding:10px 0px;"><i>meta tags</i></td>
			<td style="text-align:left;   padding:10px 0px;">net.x3djsonld.data.ScubaTank&nbsp; Document Metadata </td>
		</tr>

		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> title </i> </td>
			<td> <a href="https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Legacy/ScubaTank.x3d">ScubaTank.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> description </i> </td>
			<td> Scuba gear used by Nancy Diving example. </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> creator </i> </td>
			<td> Etsuko Lippi </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> created </i> </td>
			<td> 24 January 2001 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> modified </i> </td>
			<td> 23 May 2020 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> identifier </i> </td>
			<td> <a href="https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Legacy/ScubaTank.x3d" target="_blank">https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Legacy/ScubaTank.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> generator </i> </td>
			<td> X3D-Edit 3.3, <a href="https://savage.nps.edu/X3D-Edit" target="_blank">https://savage.nps.edu/X3D-Edit</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> license </i> </td>
			<td> <a href="https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Legacy/../license.html">../license.html</a> </td>
		</tr>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center;" colspan="2">  &nbsp; </td>
		</tr>
	</table>

	<p>
		This program uses the
		<a href="https://www.web3d.org/specifications/java/X3DJSAIL.html" target="_blank">X3D Java Scene Access Interface Library (X3DJSAIL)</a>.
		It has been produced using the 
		<a href="https://www.web3d.org/x3d/stylesheets/X3dToJava.xslt" target="_blank">X3dToJava.xslt</a>
		stylesheet
	       (<a href="https://sourceforge.net/p/x3d/code/HEAD/tree/www.web3d.org/x3d/stylesheets/X3dToJava.xslt" target="_blank">version&amp;nbsp;control</a>)
                is used to create Java source code from an original <code>.x3d</code> model.
	</p>

	* @author Etsuko Lippi
 */

public class ScubaTank
{
	/** Default constructor to create this object. */
	public ScubaTank ()
	{
	  initialize();
	}

	/** Create and initialize the X3D model for this object. */
	public final void initialize()
	{
            try { // catch-all
  x3dModel = new X3D().setProfile(X3D.PROFILE_INTERCHANGE).setVersion(X3D.VERSION_3_3)
  .setHead(new head()
    .addMeta(new meta().setName(meta.NAME_TITLE      ).setContent("ScubaTank.x3d"))
    .addMeta(new meta().setName(meta.NAME_DESCRIPTION).setContent("Scuba gear used by Nancy Diving example."))
    .addMeta(new meta().setName(meta.NAME_CREATOR    ).setContent("Etsuko Lippi"))
    .addMeta(new meta().setName(meta.NAME_CREATED    ).setContent("24 January 2001"))
    .addMeta(new meta().setName(meta.NAME_MODIFIED   ).setContent("23 May 2020"))
    .addMeta(new meta().setName(meta.NAME_IDENTIFIER ).setContent("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Legacy/ScubaTank.x3d"))
    .addMeta(new meta().setName(meta.NAME_GENERATOR  ).setContent("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"))
    .addMeta(new meta().setName(meta.NAME_LICENSE    ).setContent("../license.html")))
  .setScene(new Scene()
    .addChild(new WorldInfo().setTitle("ScubaTank.x3d"))
    .addChild(new Transform("ScubaTank")
      .addChild(new Transform()
        .addChild(new Shape()
          .setAppearance(new Appearance()
            .setMaterial(new Material("tank").setAmbientIntensity(0.3).setDiffuseColor(0.3,0.3,0.5).setShininess(0.10).setSpecularColor(0.7,0.7,0.8)))
          .setGeometry(new Cylinder().setHeight(0.7).setRadius(0.1))))
      .addChild(new Transform().setTranslation(0.0,0.35,0.0)
        .addChild(new Shape()
          .setAppearance(new Appearance()
            .setMaterial(new Material().setUSE("tank")))
          .setGeometry(new Sphere().setRadius(0.098))))
      .addChild(new Transform().setTranslation(0.0,-0.35,0.0)
        .addChild(new Shape("tankBottom")
          .setAppearance(new Appearance()
            .setMaterial(new Material("black").setAmbientIntensity(0.3).setDiffuseColor(0.0,0.0,0.0)))
          .setGeometry(new Cylinder().setHeight(0.06).setRadius(0.115))))
      .addChild(new Group("tankNozzle")
        .addChild(new Transform()
          .addChild(new Transform().setTranslation(0.0,0.45,0.0)
            .addChild(new Shape("pressure")
              .setAppearance(new Appearance()
                .setMaterial(new Material("pressureColor").setAmbientIntensity(0.4).setDiffuseColor(0.91,0.91,0.91).setShininess(0.16).setSpecularColor(0.91,0.90,0.91)))
              .setGeometry(new Cylinder().setHeight(0.1).setRadius(0.015))))
          .addChild(new Transform().setTranslation(0.0,0.50,0.0)
            .addChild(new Shape("pressureTop")
              .setAppearance(new Appearance()
                .setMaterial(new Material().setUSE("black")))
              .setGeometry(new Cylinder().setHeight(0.02).setRadius(0.025))))
          .addChild(new Transform().setRotation(0.0,0.0,1.0,1.57).setTranslation(-0.028,0.462,0.0)
            .addChild(new Transform()
              .addChild(new Shape("connectorToRegulator")
                .setAppearance(new Appearance()
                  .setMaterial(new Material().setUSE("pressureColor")))
                .setGeometry(new Cylinder().setHeight(0.03).setRadius(0.01))))
            .addChild(new Transform().setTranslation(0.0,0.02,0.0)
              .addChild(new Shape("connectorToRegulatorTop")
                .setAppearance(new Appearance()
                  .setMaterial(new Material().setUSE("black")))
                .setGeometry(new Cylinder().setHeight(0.02).setRadius(0.02)))))))
      .addChild(new Transform().setTranslation(0.0,0.2,0.0)
        .addChild(new Shape("tankHoldBelt")
          .setAppearance(new Appearance()
            .setMaterial(new Material().setUSE("black")))
          .setGeometry(new Cylinder().setHeight(0.1).setRadius(0.115)))))
    .addChild(new Background().setSkyColor(new MFColor(new double[] {0.6,0.6,0.6}))));
            }
            catch (Exception ex)
            {       
                System.err.println ("*** Further hints on X3DJSAIL errors and exceptions at");
                System.err.println ("*** https://www.web3d.org/specifications/java/X3DJSAIL.html");
                throw (ex);
            }
	}
	// end of initialize() method

	/** The initialized model object, created within initialize() method. */
	private X3D x3dModel;

	/** 
	 * Provide a 
	 * <a href="https://dzone.com/articles/java-copy-shallow-vs-deep-in-which-you-will-swim" target="_blank">shallow copy</a>
	 * of the X3D model.
	 * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/Core/X3D.html">X3D</a>
	 * @return ScubaTank model
	 */
	public X3D getX3dModel()
	{	  
		return x3dModel;
	}
	   
    /** 
	 * Default main() method provided for test purposes, uses CommandLine to set global ConfigurationProperties for this object.
     * @param args array of input parameters, provided as arguments
	 * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/Core/X3D.html#handleArguments-java.lang.String:A-">X3D.handleArguments(args)</a>
	 * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/Core/X3D.html#validationReport--">X3D.validationReport()</a>
     * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/CommandLine.html">CommandLine</a>
     * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/CommandLine.html#USAGE">CommandLine.USAGE</a>
     * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/ConfigurationProperties.html">ConfigurationProperties</a>
     */
    public static void main(String args[])
    {
        X3D thisExampleX3dModel = new ScubaTank().getX3dModel();
//      System.out.println("X3D model construction complete.");
	
        // next handle command line arguments
        boolean hasArguments = (args != null) && (args.length > 0);
        boolean validate = true; // default
        boolean argumentsLoadNewModel = false;
        String  fileName = new String();

        if (args != null)
        {
                for (String arg : args)
                {
                        if (arg.toLowerCase().startsWith("-v") || arg.toLowerCase().contains("validate"))
                        {
                                validate = true; // making sure
                        }
                        if (arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_X3D) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_CLASSICVRML) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_X3DB) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_VRML97) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_EXI) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_GZIP) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_ZIP) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_HTML) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_XHTML))
                        {
                                argumentsLoadNewModel = true;
                                fileName = arg;
                        }
                }
        }
        if      (argumentsLoadNewModel)
                System.out.println("WARNING: \"net.x3djsonld.data.ScubaTank\" model invocation is attempting to load file \"" + fileName + "\" instead of simply validating itself... file loading ignored.");
        else if (hasArguments) // if no arguments provided, this method produces usage warning
                thisExampleX3dModel.handleArguments(args);
	
        if (validate)
        {
                System.out.print("Java program \"net.x3djsonld.data.ScubaTank\" self-validation test results: ");
		String validationResults = thisExampleX3dModel.validationReport();
                if (validationResults.length() > 10)
                    System.out.println();
                System.out.println(validationResults);
        }
    }
}

File: src/main/python/net/x3djsonld/data/ScubaTank.py
####################################################################################################
#
# Invoking X3D model self-test:
#
#   $ python ScubaTank.py
#
# Python package x3d.py package is available on PyPI for import.
#   This approach simplifies Python X3D deployment and use.
#   https://pypi.org/project/x3d
#
# Installation:
#       pip install x3d
# or
#       python -m pip install x3d
#
# Developer options for loading x3d package in other Python programs:
#
#    from x3d import *  # preferred approach, terser source that avoids x3d.* class prefixes
#
# or
#    import x3d         # traditional way to subclass x3d package, all classes require x3d.* prefix,
#                       # but python source is very verbose, for example x3d.Material x3d.Shape etc.
#                       # X3dToPython.xslt stylesheet insertPackagePrefix=true supports this option.
#
####################################################################################################

from x3d import *

newModel=X3D(profile='Interchange',version='4.0',
  head=head(
    children=[
    meta(content='ScubaTank.x3d',name='title'),
    meta(content='Scuba gear used by Nancy Diving example.',name='description'),
    meta(content='Etsuko Lippi',name='creator'),
    meta(content='24 January 2001',name='created'),
    meta(content='23 May 2020',name='modified'),
    meta(content='https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Legacy/ScubaTank.x3d',name='identifier'),
    meta(content='X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit',name='generator'),
    meta(content='../license.html',name='license')]),
  Scene=Scene(
    children=[
    WorldInfo(title='ScubaTank.x3d'),
    Transform(DEF='ScubaTank',
      children=[
      Transform(
        children=[
        Shape(
          appearance=Appearance(
            material=Material(DEF='tank',ambientIntensity=0.3,diffuseColor=(0.3,0.3,0.5),shininess=0.10,specularColor=(0.7,0.7,0.8))),
          geometry=Cylinder(height=0.7,radius=0.1))]),
      Transform(translation=(0,0.35,0),
        children=[
        Shape(
          appearance=Appearance(
            material=Material(USE='tank')),
          geometry=Sphere(radius=0.098))]),
      Transform(translation=(0,-0.35,0),
        children=[
        Shape(DEF='tankBottom',
          appearance=Appearance(
            material=Material(DEF='black',ambientIntensity=0.3,diffuseColor=(0,0,0))),
          geometry=Cylinder(height=0.06,radius=0.115))]),
      Group(DEF='tankNozzle',
        children=[
        Transform(
          children=[
          Transform(translation=(0,0.45,0),
            children=[
            Shape(DEF='pressure',
              appearance=Appearance(
                material=Material(DEF='pressureColor',ambientIntensity=0.4,diffuseColor=(0.91,0.91,0.91),shininess=0.16,specularColor=(0.91,0.90,0.91))),
              geometry=Cylinder(height=0.1,radius=0.015))]),
          Transform(translation=(0,0.50,0),
            children=[
            Shape(DEF='pressureTop',
              appearance=Appearance(
                material=Material(USE='black')),
              geometry=Cylinder(height=0.02,radius=0.025))]),
          Transform(rotation=(0,0,1,1.57),translation=(-0.028,0.462,0),
            children=[
            Transform(
              children=[
              Shape(DEF='connectorToRegulator',
                appearance=Appearance(
                  material=Material(USE='pressureColor')),
                geometry=Cylinder(height=0.03,radius=0.01))]),
            Transform(translation=(0,0.02,0),
              children=[
              Shape(DEF='connectorToRegulatorTop',
                appearance=Appearance(
                  material=Material(USE='black')),
                geometry=Cylinder(height=0.02,radius=0.02))])])])]),
      Transform(translation=(0,0.2,0),
        children=[
        Shape(DEF='tankHoldBelt',
          appearance=Appearance(
            material=Material(USE='black')),
          geometry=Cylinder(height=0.1,radius=0.115))])]),
    Background(skyColor=[(0.6,0.6,0.6)])])
)
### X3D model conversion complete ###

####################################################################################################
# Self-test diagnostics
####################################################################################################

print('Self-test diagnostics for ScubaTank.py:')
if        metaDiagnostics(newModel): # built-in utility method in X3D class
    print(metaDiagnostics(newModel)) # display meta info, hint, warning, error, TODO values in this model
# print('check newModel.XML() serialization...')
newModelXML= newModel.XML() # test export method XML() for exceptions during export
newModel.XMLvalidate()
# print(newModelXML) # diagnostic

try:
#   print('check newModel.VRML() serialization...')
    newModelVRML=newModel.VRML() # test export method VRML() for exceptions during export
    # print(prependLineNumbers(newModelVRML)) # debug
    print("Python-to-VRML export of VRML output successful", flush=True)
except Exception as err: # usually BaseException
    # https://stackoverflow.com/questions/18176602/how-to-get-the-name-of-an-exception-that-was-caught-in-python
    print("*** Python-to-VRML export of VRML output failed:", type(err).__name__, err)
    if newModelVRML: # may have failed to generate
        print(prependLineNumbers(newModelVRML, err.lineno))

try:
#   print('check newModel.JSON() serialization...')
    newModelJSON=newModel.JSON() # test export method JSON() for exceptions during export
#   print(prependLineNumbers(newModelJSON)) # debug
    print("Python-to-JSON export of JSON output successful (under development)")
except Exception as err: # usually SyntaxError
    print("*** Python-to-JSON export of JSON output failed:", type(err).__name__, err)
    if newModelJSON: # may have failed to generate
        print(prependLineNumbers(newModelJSON,err.lineno))

print("python ScubaTank.py load and self-test diagnostics complete.")
